File: csharp/APITest/APITest/ZelloAPI.cs
//
//  Copyright © 2016 Zello. All rights reserved.
//

using System;
using System.Threading.Tasks;
using System.Net;
using System.Text;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Web.Script.Serialization;
using System.Security.Cryptography;

namespace APITest
{
    /// <summary>
    /// ZelloWork server C# API wrapper class.
    /// This class provides an easy way to interact with the ZelloWork server
    /// from your C# code to add, modify and delete users and channels.
    /// Please note that all text values passed to the API must be in UTF-8 encoding
    /// and any text data returned are in UTF-8 as well.
    /// - Version 1.1
    /// </summary>
    public class ZelloAPI
    {
        /// <summary>
        /// API Version
        /// </summary>
        public static string Version = "1.1";

        /// <summary>
        /// Session ID used to identify logged in client. Typically you'll want to authenticate first and store the Session ID to reuse later.
        /// </summary>
        public string SessionId;

        /// <summary>
        /// Last accessed API URL. Useful for API troubleshooting.
        /// </summary>
        public string LastURL;

        /// <summary>
        /// Server hostname or IP address
        /// </summary>
        string host;
        /// <summary>
        /// API Key
        /// </summary>
        string apiKey;

        /// <summary>
        /// Initializes a new instance of the <see cref="T:Zello.API.ZelloAPI"/> class.
        /// </summary>
        /// <param name="host">Host.</param>
        /// <param name="apiKey">API key.</param>
        public ZelloAPI(string host, string apiKey)
            : this(host, apiKey, null)
        {

        }

        /// <summary>
        /// Initializes a new instance of the <see cref="T:Zello.API.ZelloAPI"/> class.
        /// </summary>
        /// <param name="host">Host.</param>
        /// <param name="apiKey">API key.</param>
        /// <param name="sessionId">Optional session identifier. Can be null.</param>
        public ZelloAPI(string host, string apiKey, string sessionId)
        {
            this.host = host;
            this.apiKey = apiKey;
            SessionId = sessionId;
        }

        /// <summary>
        /// API client authentication.
        /// If authentication fails, use the errorCode and errorDescription attributes on the response dictionary to get error details.
        /// If authentication succeeds, SessionId is set to the Session ID.
        /// The Session ID is reusable so it's recommended that you save this value and use it for further API calls.
        /// Once you are done using the API, call ZelloAPI.Logout() to end the session and invalidate Session ID.
        /// </summary>
        /// <param name="username">administrative username.</param>
        /// <param name="password">administrative password.</param>
        public async Task<ZelloAPIResult> Authenticate(string username, string password)
        {
            ZelloAPIResult result = await callAPI("user/gettoken", HTTPMethod.GET, null).ConfigureAwait(false);

            ZelloAPIResult returnResult;
            if (!result.Success)
            {
                returnResult = new ZelloAPIResult(false, result.Response, null);
                return returnResult;
            }

            if (result.Response == null)
            {
                returnResult = new ZelloAPIResult(false, null, null);
                return returnResult;
            }

            var token = (string)result.Response["token"];

            SessionId = (string)result.Response["sid"];

            if (apiKey == null)
            {
                returnResult = new ZelloAPIResult(false, result.Response, null);
                return returnResult;
            }

            string hashedPassword = MD5Hash(MD5Hash(password) + token + apiKey);
            string parameters = "username=" + username + "&password=" + hashedPassword;

            returnResult = await callAPI("user/login", HTTPMethod.POST, parameters);

            return returnResult;
        }

        /// <summary>
        /// Ends session identified by sessionId.
        /// Use this method to terminate the API session.
        /// See ZelloAPI.Authenticate().
        /// </summary>
        public async Task<ZelloAPIResult> Logout()
        {
            ZelloAPIResult result = await callAPI("user/logout", HTTPMethod.GET, null);
            SessionId = null;

            return result;
        }

        /// <summary>
        /// Gets the list of the users or detailed information regarding a particular user.
        /// </summary>
        /// <param name="username">username of the user, for which the details are requested. If null, the full users list is returned.</param>
        /// <param name="isGateway">whether to return users or gateways.</param>
        /// <param name="max">maximum number of results to fetch.</param>
        /// <param name="start">start index of results to fetch.</param>
        /// <param name="channel">channel name.</param>
        public async Task<ZelloAPIResult> GetUsers(string username, bool isGateway, int? max, int? start, string channel)
        {
            string command = "user/get";

            if (username != null)
            {
                command += "/login/" + urlEncode(username);
            }
            if (channel != null)
            {
                command += "/channel/" + urlEncode(channel);
            }
            if (isGateway)
            {
                command += "/gateway/1";
            }
            if (max != null)
            {
                command += "/max/" + max;
            }
            if (start != null)
            {
                command += "/start/" + start;
            }

            return await callAPI(command, HTTPMethod.GET, null);
        }

        /// <summary>
        /// Gets the list of the channels or detailed information regarding a particular channel.
        /// </summary>
        /// <param name="name">optional name of the channel, for which the details are requested. If omitted, the full channels list is returned.</param>
        /// <param name="max">maximum number of results to fetch.</param>
        /// <param name="start">start index of results to fetch.</param>
        public async Task<ZelloAPIResult> GetChannels(string name, int? max, int? start)
        {
            string command = "channel/get";

            if (name != null)
            {
                command += "/name/" + urlEncode(name);
            }
            if (max != null)
            {
                command += "/max/" + max;
            }
            if (start != null)
            {
                command += "/start/" + start;
            }

            return await callAPI(command, HTTPMethod.GET, null);
        }

        /// <summary>
        /// Adds users to a channel.
        /// </summary>
        /// <param name="channelName">name of the channel where the users will be added.</param>
        /// <param name="users">usernames of the users to add.</param>
        public async Task<ZelloAPIResult> AddToChannel(string channelName, ArrayList users)
        {
            string command = "user/addto/" + urlEncode(channelName);

            string parameters = implode("login[]=", "&login[]=", users);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Add users to multiple channels.
        /// </summary>
        /// <param name="channelNames">channel names where the users are added.</param>
        /// <param name="users">usernames of the users to add.</param>
        public async Task<ZelloAPIResult> AddToChannels(ArrayList channelNames, ArrayList users)
        {
            string command = "user/addtochannels";

            string parameters = implode("users[]=", "&users[]=", users);
            parameters += implode("&channels[]=", "&channels[]=", channelNames);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Removes users from a channel.
        /// </summary>
        /// <param name="channelName">name of the channel.</param>
        /// <param name="users">usernames of the users to remove from the channel.</param>
        public async Task<ZelloAPIResult> RemoveFromChannel(string channelName, ArrayList users)
        {
            string command = "user/removefrom/" + urlEncode(channelName);

            string parameters = implode("login[]=", "&login[]=", users);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Removes users from multiple channels.
        /// </summary>
        /// <param name="channelNames">names of the channels.</param>
        /// <param name="users">usernames of the users to remove from channels.</param>
        public async Task<ZelloAPIResult> RemoveFromChannels(ArrayList channelNames, ArrayList users)
        {
            string command = "user/removefromchannels";

            string parameters = implode("users[]=", "&users[]=", users);
            parameters += implode("&channels[]=", "&channels[]=", channelNames);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Adds or updates the user.
        /// If username exists, the user is updated.Otherwise, a new user is created.
        /// When adding a user, the "name" and "password" attributes are required.
        /// When updating a user, the "name" attribute is required.
        ///
        /// See ZelloAPI.deleteUsers()
        /// 
        /// Attributes:
        /// - name(required) - username
        /// - password - password md5 hash
        /// - email - e-mail address
        /// - full_name - user alias
        /// - job - user position
        /// - admin - "true" or "false". Defines whether the user has access to the admin console.
        /// - limited_access - "true" or "false". Defines whether the user is restricted from starting 1-on-1 conversations or not.
        /// - gateway - set to "true" for adding a gateway, "false" if normal user.
        /// - add - "true" or "false". If set to "true", the existing user will not be updated and an error will be returned.		
        /// </summary>
        /// <param name="user">dictionary of user attributes.</param>
        public async Task<ZelloAPIResult> SaveUser(Dictionary<string, string> user)
        {
            string command = "user/save";

            string parameters = createURLStringFromDictionary(user);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Deletes users.
        ///
        /// See ZelloAPI.saveUser().
        /// </summary>
        /// <param name="users">usernames of the users to remove.</param>
        public async Task<ZelloAPIResult> DeleteUsers(ArrayList users)
        {
            string command = "user/delete";

            string parameters = implode("login[]=", "&login[]=", users);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Adds a new channel.
        /// </summary>
        /// <param name="name">channel name.</param>
        /// <param name="isGroup">true means it is a group channel. false means it is a dynamic channel.</param>
        /// <param name="isHidden">when set to true in combination with isGroup set to true, a hidden group channel is created.</param>
        public async Task<ZelloAPIResult> AddChannel(string name, bool? isGroup, bool? isHidden)
        {
            string command = "channel/add/name/" + urlEncode(name);

            if (isGroup != null)
            {
                command += "/shared/" + ((bool)isGroup ? "true" : "false");
            }
            if (isHidden != null)
            {
                command += "/invisible/" + ((bool)isHidden ? "true" : "false");
            }

            return await callAPI(command, HTTPMethod.GET, null);
        }

        /// <summary>
        /// Deletes channels.
        /// </summary>
        /// <param name="channelNames">names of the channels to remove.</param>
        public async Task<ZelloAPIResult> DeleteChannels(ArrayList channelNames)
        {
            string command = "channel/delete";

            string parameters = implode("name[]=", "&name[]=", channelNames);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Get channel roles (simple format).
        /// </summary>
        /// <param name="channelName">channel name.</param>
        public async Task<ZelloAPIResult> GetChannelsRoles(string channelName)
        {
            string command = "channel/roleslist/name/" + urlEncode(channelName);

            return await callAPI(command, HTTPMethod.GET, null);
        }

        /// <summary>
        /// Adds or updates channel role.
        /// </summary>
        /// <param name="channelName">channel name.</param>
        /// <param name="roleName">new role name.</param>
        /// <param name="settings">role settings in JSON format: ["listen_only" : false, "no_disconnect" : true, "allow_alerts" : false, "to": ["dispatchers"]].</param>
        public async Task<ZelloAPIResult> SaveChannelRole(string channelName, string roleName, Dictionary<string, object> settings)
        {
            string command = "channel/saverole/channel/" + urlEncode(channelName) + "/name/" + urlEncode(roleName);

            string json = new JavaScriptSerializer().Serialize(settings);
            string parameters = "settings=" + json;

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Deletes channel role.
        /// </summary>
        /// <param name="channelName">channel name.</param>
        /// <param name="roles">role names to delete.</param>
        public async Task<ZelloAPIResult> DeleteChannelRole(string channelName, ArrayList roles)
        {
            string command = "channel/deleterole/channel/" + urlEncode(channelName);

            string parameters = implode("roles[]=", "&roles[]=", roles);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        /// <summary>
        /// Adds users to role in channel.
        /// </summary>
        /// <param name="channelName">channel name.</param>
        /// <param name="roleName">role name.</param>
        /// <param name="users">usernames to add to role in channel.</param>
        public async Task<ZelloAPIResult> AddToChannelRole(string channelName, string roleName, ArrayList users)
        {
            string command = "channel/addtorole/channel/" + urlEncode(channelName) + "/name/" + urlEncode(roleName);

            string parameters = implode("login[]=", "&login[]=", users);

            return await callAPI(command, HTTPMethod.POST, parameters);
        }

        async Task<ZelloAPIResult> callAPI(string command, HTTPMethod method, string parameters)
        {
            ZelloAPIResult returnResult;

            string prefix = "http://";
            if (host.Contains("http://") || host.Contains("https://"))
            {
                prefix = "";
            }
            string urlString = prefix + host + "/" + command;

            if (SessionId != null)
            {
                urlString += "?sid=" + SessionId;
            }

            LastURL = urlString;

            var request = (HttpWebRequest)WebRequest.Create(urlString);
            request.Method = convertHTTPMethodToString(method);

            if (parameters != null)
            {
                request.ContentType = "application/x-www-form-urlencoded; charset=utf-8";

                byte[] postData = Encoding.UTF8.GetBytes(parameters);

                using (var requestStream = await Task<Stream>.Factory.FromAsync(request.BeginGetRequestStream, request.EndGetRequestStream, request))
                {
                    requestStream.Write(postData, 0, postData.Length);
                }
            }

            try
            {
                // Pick up the response:
                using (var response = (HttpWebResponse)(await Task<WebResponse>.Factory.FromAsync(request.BeginGetResponse, request.EndGetResponse, null)))
                {
                    var reader = new StreamReader(response.GetResponseStream());
                    string result = reader.ReadToEnd();
                    var json = (Dictionary<string, object>)new JavaScriptSerializer().DeserializeObject(result);

                    bool success = json != null && json["code"].Equals("200");
                    returnResult = new ZelloAPIResult(success, json, null);
                }
            }
            catch (WebException exception)
            {
                returnResult = new ZelloAPIResult(false, null, exception);
            }

            return returnResult;
        }

        string convertHTTPMethodToString(HTTPMethod method)
        {
            switch (method)
            {
                case HTTPMethod.GET:
                    return "GET";
                case HTTPMethod.POST:
                    return "POST";
            }

            return "";
        }

        string MD5Hash(string input)
        {
            // Convert the input string to a byte array and compute the hash.
            byte[] data = MD5.Create().ComputeHash(Encoding.UTF8.GetBytes(input));

            // Create a new StringBuilder to collect the bytes
            // and create a string.
            var sBuilder = new StringBuilder();

            // Loop through each byte of the hashed data 
            // and format each one as a hexadecimal string.
            for (int i = 0; i < data.Length; i++)
            {
                sBuilder.Append(data[i].ToString("x2"));
            }

            // Return the hexadecimal string.
            return sBuilder.ToString();
        }

        string urlEncode(string input)
        {
            return WebUtility.UrlEncode(input);
        }

        string implode(string s, string glue, ArrayList pieces)
        {
            for (int i = 0; i < pieces.Count; i++)
            {
                s += urlEncode((string)pieces[i]);

                if (i < pieces.Count - 1)
                {
                    s += glue;
                }
            }

            return s;
        }

        string createURLStringFromDictionary(Dictionary<string, string> dictionary)
        {
            string s = "";

            int i = 0;
            foreach (KeyValuePair<string, string> entry in dictionary)
            {
                string key = entry.Key;
                string value = entry.Value;

                if (i == 0)
                {
                    s += key + "=" + urlEncode(value);
                }
                else
                {
                    s += "&" + key + "=" + urlEncode(value);
                }

                i++;
            }

            return s;
        }
    }

    /// <summary>
    /// Completion Object for making API Requests that returns a success indicator, a response dictionary and an exception.
    /// A value for the exception variable represents a client error. This is not the error returned by the API.
    /// An error returned by the API can be retrieved through the response dictionary.
    /// </summary>
    public class ZelloAPIResult
    {
        public bool Success;
        public Dictionary<string, object> Response;
        public Exception Exception;

        public ZelloAPIResult(bool success, Dictionary<string, object> response, Exception exception)
        {
            Success = success;
            Response = response;
            Exception = exception;
        }
    }

    enum HTTPMethod
    {
        GET, POST
    }
}

File: swift/CocoaPod/ZelloAPISwift/ZelloAPI.swift
//
//  Copyright © 2016 Zello. All rights reserved.
//

import Foundation
import CommonCrypto

/// Completion Handler for making API Requests that returns a success indicator, result dictionary and error.
/// A value for the error variable represents a client error. This is not the error returned by the API.
/// An error returned by the API can be retrieved through the result dictionary.
/// This closure is always executed on the main thread.
public typealias ResultCompletionHandler = (Bool, [String : AnyObject]?, NSError?) -> Void

/**
 ZelloWork server Swift API wrapper class.

 This class provides an easy way to interact with the ZelloWork server
 from your Swift code to add, modify and delete users and channels.
 Please note that all text values passed to the API must be in UTF-8 encoding
 and any text data returned are in UTF-8 as well.

 - Version 1.1.0
 - Swift Version 3.0
 - Minimum iOS Version 8.0
*/
open class ZelloAPI {
  
  // MARK: Public Variables
  
  /// API Version
  public static let version = "1.1.0"
  
  /// Session ID used to identify logged in client. Typically you'll want to authenticate first and store the Session ID to reuse later.
  open var sessionId: String?
  
  /// Last accessed API URL. Useful for API troubleshooting.
  open var lastURL: String?
  
  // MARK: Private Variables
  
  /// Server hostname or IP address
  fileprivate var host: String?
  /// API Key
  fileprivate var apiKey: String?
  
  // MARK: Initializer
  
  public convenience init(host: String, apiKey: String, sessionId: String?=nil) {
    self.init()
    
    self.host = host
    self.apiKey = apiKey
    self.sessionId = sessionId
  }
  
  // MARK: Public Methods
  
  /**
   API client authentication.
   If authentication fails, use the errorCode and errorDescription attributes on the response dictionary to get error details.
   If authentication succeeds, sessionId is set to the Session ID.
   The Session ID is reusable so it's recommended that you save this value and use it for further API calls.
   Once you are done using the API, call ZelloAPI.logout() to end the session and invalidate Session ID.
   
   - parameter username:          administrative username
   - parameter password:          administrative password
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func authenticate(_ username: String, password: String, completionHandler: @escaping ResultCompletionHandler) {
    callAPI("user/gettoken", httpMethod: .GET, completionHandler: { [weak self] (success, response, error) -> Void in
      // On main thread
      guard let weakSelf = self else {
        completionHandler(false, response, error)
        return
      }
      
      if !success {
        completionHandler(success, response, error)
        return
      }
      
      guard let response = response else {
        completionHandler(false, nil, error)
        return
      }
      
      guard let token = response["token"] as? String else {
        completionHandler(false, response, error)
        return
      }
      
      weakSelf.sessionId = response["sid"] as? String
      
      guard let apiKey = weakSelf.apiKey else {
        completionHandler(false, response, error)
        return
      }
      
      let parameters = "username=\(username)&password=\((password.MD5() + token + apiKey).MD5())"
      
      weakSelf.callAPI("user/login", httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
    })
  }
  
  /**
   Ends session identified by sessionId.
   Use this method to terminate the API session.
   See ZelloAPI.authenticate()
   
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func logout(_ completionHandler: @escaping ResultCompletionHandler) {
    callAPI("user/logout", httpMethod: .GET, completionHandler: { [weak self] (success, response, error) -> Void in
      self?.sessionId = nil
      
      completionHandler(success, response, error)
    })
  }
  
  /**
   Gets the list of the users or detailed informatino regarding a particular user.
   
   - parameter username:          username of the user, for which the details are requested. If nil, the full users list is returned.
   - parameter isGateway:         whether to return users or gateways.
   - parameter max:               maximum number of results to fetch.
   - parameter start:             start index of results to fetch.
   - parameter channel:           channel name.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func getUsers(_ username: String?=nil, isGateway: Bool?=false, max: Int?=nil, start: Int?=nil, channel: String?=nil, completionHandler: @escaping ResultCompletionHandler) {
    var command = "user/get"
    
    if let username = username {
      command += "/login/" + username.urlEncode()
    }
    if let channel = channel {
      command += "/channel/" + channel.urlEncode()
    }
    if let isGateway = isGateway , isGateway {
      command += "/gateway/1"
    }
    if let max = max {
      command += "/max/" + String(max)
    }
    if let start = start {
      command += "/start/" + String(start)
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Gets the list of the channels or detailed information regarding a particular channel.
   
   - parameter name:              name of the channel, for which the details are requested. If omitted, the full channels list is returned.
   - parameter max:               maximum number of results to fetch.
   - parameter start:             start index of results to fetch.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func getChannels(_ name: String?=nil, max: Int?=nil, start: Int?=nil, completionHandler: @escaping ResultCompletionHandler) {
    var command = "channel/get"
    
    if let name = name {
      command += "/name/" + name.urlEncode()
    }
    if let max = max {
      command += "/max/" + String(max)
    }
    if let start = start {
      command += "/start/" + String(start)
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Adds users to a channel.
   
   - parameter channelName: name of the channel where the users will be added.
   - parameter users:       usernames of the users to add.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func addToChannel(_ channelName: String, users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/addto/" + channelName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)

    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Add users to multiple channels.
   
   - parameter channelNames:      channel names where the users are added.
   - parameter users:             usernames of the users to add.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func addToChannels(_ channelNames: [String], users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/addtochannels"
    
    var parameters = "users[]=".implode("&users[]=", pieces: users)
    parameters += "&channels[]=".implode("&channels[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Removes users from a channel.
   
   - parameter channelName:       name of the channel.
   - parameter users:             usernames of the users to remove from the channel.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func removeFromChannel(_ channelName: String, users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/removefrom/" + channelName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Removes users from multiple channels.
   
   - parameter channelNames:      names of the channels.
   - parameter users:             usernames of the users to remove from channels.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func removeFromChannels(_ channelNames: [String], users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/removefromchannels"
    
    var parameters = "users[]=".implode("&users[]=", pieces: users)
    parameters += "&channels[]=".implode("&channels[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds or updates the user.
   If username exists, the user is updated. Otherwise, a new user is created.
   When adding a user, the "name" and "password" attributes are required.
   When updating a user, the "name" attribute is required.
   
   See ZelloAPI.deleteUsers()
   
   Attributes:
   - name (required) - username
   - password - password md5 hash
   - email - e-mail address
   - full_name - user alias
   - job - user position
   - admin - "true" or "false". Defines whether the user has access to the admin console.
   - limited_access - "true" or "false". Defines whether the user is restricted from starting 1-on-1 conversations or not.
   - gateway - set to "true" for adding a gateway, "false" if normal user.
   - add - "true" or "false". If set to "true", the existing user will not be updated and an error will be returned.
   
   - parameter user:              dictionary of user attributes.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func saveUser(_ user: [String : String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/save"
    
    let parameters = createURLStringFromDictionary(user)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Deletes users.
   
   See ZelloAPI.saveUser()
   
   - parameter users:             usernames of the users to remove.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func deleteUsers(_ users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "user/delete"
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds a new channel.
   
   - parameter name:              channel name.
   - parameter isGroup:           true means it is a group channel. false means it is a dynamic channel.
   - parameter isHidden:          when set to true in combination with isGroup set to true, a hidden group channel is created.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func addChannel(_ name: String, isGroup: Bool?=nil, isHidden: Bool?=nil, completionHandler: @escaping ResultCompletionHandler) {
    var command = "channel/add/name/" + name.urlEncode()
    
    if let isGroup = isGroup {
      command += "/shared/" + (isGroup ? "true" : "false")
    }
    if let isHidden = isHidden {
      command += "/invisible/" + (isHidden ? "true" : "false")
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Deletes channels.
   
   - parameter channelNames:      names of the channels to remove.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func deleteChannels(_ channelNames: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "channel/delete"
    
    let parameters = "name[]=".implode("&name[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Get channel roles (simple format).
   
   - parameter channelName:       channel name.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func getChannelsRoles(_ channelName: String, completionHandler: @escaping ResultCompletionHandler) {
    let command = "channel/roleslist/name/" + channelName.urlEncode()
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Adds or updates channel role.
   
   - parameter channelName:       channel name.
   - parameter roleName:          new role name.
   - parameter settings:          role settings in JSON format: ["listen_only" : false, "no_disconnect" : true, "allow_alerts" : false, "to": ["dispatchers"]]
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func saveChannelRole(_ channelName: String, roleName: String, settings: [String : AnyObject], completionHandler: @escaping ResultCompletionHandler) {
    let command = "channel/saverole/channel/" + channelName.urlEncode() + "/name/" + roleName.urlEncode()
    
    var parameters: String?
    do {
      let data = try JSONSerialization.data(withJSONObject: settings, options: JSONSerialization.WritingOptions(rawValue: 0))
      if let dataString = String(data: data, encoding: String.Encoding.utf8) {
        parameters = "settings=" + dataString
      }
    } catch {
      
    }
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Deletes channel role.
   
   - parameter channelName:       channel name.
   - parameter roles:             role names to delete.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func deleteChannelRole(_ channelName: String, roles: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "channel/deleterole/channel/" + channelName.urlEncode()
    
    let parameters = "roles[]=".implode("&roles[]=", pieces: roles)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds users to role in channel.
   
   - parameter channelName:       channel name.
   - parameter roleName:          role name.
   - parameter users:             usernames to add to role in channel.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  open func addToChannelRole(_ channelName: String, roleName: String, users: [String], completionHandler: @escaping ResultCompletionHandler) {
    let command = "channel/addtorole/channel/" + channelName.urlEncode() + "/name/" + roleName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  // MARK: Private Methods
  
  fileprivate func callAPI(_ command: String, httpMethod: HTTPMethod, parameters: String?=nil, completionHandler: @escaping ResultCompletionHandler) {
    let session = URLSession.shared
    
    guard let host = host else {
      return
    }
    
    var prefix = "http://"
    if host.contains("http://") || host.contains("https://") {
      prefix = ""
    }
    
    var urlString = prefix + host + "/" + command

    if let sessionId = sessionId {
      urlString += "?sid=" + sessionId
    }
    
    lastURL = urlString
    
    guard let nsURL = URL(string: urlString) else {
      DispatchQueue.main.async {
        completionHandler(false, nil, NSError.unknownError())
      }
      return
    }
    
    let request = NSMutableURLRequest(url: nsURL as URL)
    request.httpMethod = httpMethod.rawValue
    
    if let parameters = parameters {
      request.setBodyContentFromString(parameters)
    }
    
    session.dataTask(with: request as URLRequest) { (data, response, error) -> Void in
      if error == nil {
        guard let data = data else {
          DispatchQueue.main.async {
            completionHandler(false, nil, NSError.unknownError())
          }
          return
        }
        
        var responseDictionary: [String : AnyObject]?
        do {
          responseDictionary = try JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions(rawValue: 0)) as? [String : AnyObject]
        } catch let error as NSError {
          DispatchQueue.main.async {
            completionHandler(false, nil, error)
          }
          return
        }
        
        guard let unwrappedResponseDictionary = responseDictionary else {
          DispatchQueue.main.async {
            completionHandler(false, nil, NSError.unknownError())
          }
          return
        }
        
        guard let statusCode = unwrappedResponseDictionary["code"] as? String else {
          DispatchQueue.main.async {
            completionHandler(false, unwrappedResponseDictionary, NSError.unknownError())
          }
          return
        }
        
        let success = statusCode == "200"
        DispatchQueue.main.async {
          completionHandler(success, unwrappedResponseDictionary, nil)
        }
        return
      } else {
        DispatchQueue.main.async {
          completionHandler(false, nil, NSError.unknownError())
        }
        return
      }
    }.resume()
  }
  
  fileprivate func createURLStringFromDictionary(_ dictionary: [String : String]) -> String {
    var string = ""
    for (index, key) in dictionary.keys.enumerated() {
      if index == 0 {
        string += key + "=" + dictionary[key]!.urlEncode()
      } else {
        string += "&" + key + "=" + dictionary[key]!.urlEncode()
      }
    }
    
    return string
  }

}

// MARK: Helpers

/**
 Represents the different HTTP methods that can be used when making requests to the Zello Server API.
 */
private enum HTTPMethod: String {
  
  case POST = "POST"
  case GET = "GET"
  
}

private enum ZelloAPIErrorDomain: String {
  
  case UnknownError = "ZELLOUnknownError"

}

private extension NSError {
  
  static func unknownError() -> NSError {
    return NSError(domain: ZelloAPIErrorDomain.UnknownError.rawValue, code: 0, userInfo: nil)
  }
  
}

private extension String {
  
  /// Calculates the MD5 hash of a string.
  func MD5() -> String {
    guard let str = cString(using: .utf8) else {
      return ""
    }
    
    let strLen = CC_LONG(lengthOfBytes(using: String.Encoding.utf8))
    let digestLen = Int(CC_MD5_DIGEST_LENGTH)
    let result = UnsafeMutablePointer<CUnsignedChar>.allocate(capacity: digestLen)
    
    CC_MD5(str, strLen, result)
    
    let hash = NSMutableString()
    for i in 0..<digestLen {
      hash.appendFormat("%02x", result[i])
    }
    
    result.deallocate()
    
    return String(format: hash as String)
  }
  
  /// URL encodes a string.
  func urlEncode() -> String {
    if let string = addingPercentEncoding(withAllowedCharacters: CharacterSet.alphanumerics) {
      return string
    }
    
    return self
  }

  /**
   Join array elements with a glue string. URL encodes all pieces.
   */
  func implode(_ glue: String, pieces: [String]) -> String {
    var string = self
    
    for (index, piece) in pieces.enumerated() {
      string += piece.urlEncode()
      
      if (index < pieces.count - 1) {
        string += glue
      }
    }
    
    return string
  }
  
}

private extension NSMutableURLRequest {
  
  /**
   Sets the HTTPBody of the NSMutableURLRequest to the contents of a string.
   
   - parameter string: string to encode into HTTPBody.
   */
  func setBodyContentFromString(_ string: String) {
    httpBody = string.data(using: String.Encoding.utf8)
    
    addValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")
  }
  
}
